### PR TITLE
Refactor proposals

### PR DESCRIPTION
**Pull request by fsmanuel**
@skddc can you help me to understand how the queryParams are used? I removed them for now (see the commit message) but I'm happy to readd them.

### CODE REVIEW

File: app/controllers/proposals/new.js [hunk 1]
```diff
@@ -1,29 +1,24 @@
...
 
-  contributors: null,
+  contributors: [],
+  minedContributors: filterBy('contributors', 'id'),
 
   actions: {
-    onSave() {
-      this.transitionToRoute('index');
+    save(proposal) {
+      // contributorIpfsHash is needed for the proposal ipfs data. I'm not happy to do this here but I think to load all the contributors in addProposal again is a bit too much. I hope we can refactor it later.
```
Comment (fsmanuel) [inline app/controllers/proposals/new.js]: @bumi we need to look into hubot and how we should do the contributor look up there. I think if we have some kind of caching it would be better to have it in the kredits module

Comment (bumi) [reply app/controllers/proposals/new.js]: I'd remove the contributor ipfsHash from the proposal and only have the kredits id stored there. With that one the latest profile ipfs hash can be retrieved from the contract. This would also make sure that one always gets the latest profile hash of the contributor. 

hope I did not forget some reason why the hash is stored there.

Comment (skddc) [reply app/controllers/proposals/new.js]: It is stored there, because you need to be able to access the profile at the time of the proposal.

**Comment by skddc** [comment]
> @skddc can you help me to understand how the queryParams are used? I removed them for now (see the commit message) but I'm happy to readd them.

They are used for linking a new proposal creation form, with any of the fields pre-set by the URL.

File: app/services/kredits.js [hunk 2]
```diff
@@ -115,6 +118,14 @@ export default Service.extend({
     return contributor;
   },
 
+  buildProposal(attributes) {
+    debug('[kredits] buildProposal', attributes);
+
+    let proposal = getOwner(this).lookup('model:proposal');
+    proposal.setProperties(attributes);
+    return proposal;
```
Comment (skddc) [inline app/services/kredits.js]: This looks a bit strange to me. Is there a benefit to not just importing the model and doing a `create()` or something? (I'm not actually sure what's really happening here though.)

File: app/services/kredits.js [hunk 3]
```diff
@@ -283,29 +300,40 @@ export default Service.extend({
...
+
+    let json = ContributionSerializer.serialize(attributes);
+    // TODO: validate against schema
 
     return this.get('ipfs')
-      .storeFile(proposal.serializeContribution())
-      .then(ipfsHash => {
+      .storeFile(json)
+      // Set ipfsHash
+      .then((ipfsHash) => {
```
Comment (skddc) [inline app/services/kredits.js]: Just a formatting thing: if it's only one function argument, we don't need the parentheses around it. So it looks a bit cleaner and reads just a tiny little bit easier.

File: app/services/kredits.js [hunk 4]
```diff
@@ -283,29 +300,40 @@ export default Service.extend({
...
-              url,
-              ipfsHash
-            );
-          })
-          .then((data) => {
-            debug('[kredits] add proposal response', data);
-            return data;
+              digest,
+              hashFunction,
+              hashSize,
```
Comment (skddc) [inline app/services/kredits.js]: I think the hash splitting and combining should not be done in here, but needs to in our kredits npm module, so we can use it from other programs as well.

File: tests/unit/models/proposal-test.js [hunk 5]
```diff
@@ -1,49 +1,8 @@
...
-    recipientProfile: 'QmT2A7rY4e7uoKktkcFHQNN7BD1oXdZTgd8wNkr1u9nNVE'
-  });
-
-  try {
-    let json = model.contributionToJSON();
-    assert.notOk(json, true);
-  } catch(e) {
-    assert.ok(e.message.match(/Missing .* kind.*description/i));
-  }
-});
```
Comment (skddc) [inline tests/unit/models/proposal-test.js]: The validations seem to be missing in the new serializer tests.

File: app/services/kredits.js (hunk 2 above)
Comment (fsmanuel) [reply app/services/kredits.js]: The problem with the `.create()` is that we can not inject services because the model doesn't have a container. I'm unhappy with this as well and will try to figure out a better way.

File: app/services/kredits.js (hunk 4 above)
Comment (fsmanuel) [reply app/services/kredits.js]: What we try right now is that we cleanup the kredits service and and extract all the common logic from the service in the npm module. That will be our next step.

File: app/controllers/proposals/new.js (hunk 1 above)
Comment (bumi) [reply app/controllers/proposals/new.js]: hmmm. but all data is stored in the contract and that one holds the "truth". 
the proposal IPFS hash is retrieved from the contract and through the contracts the contributor ipfs hash can be retrieved. (the latest one, but also thanks to blockchain the one at a specific time in the past - when the proposal was created). 

To load data might be one more step: get the contributor by id from the contract and then access the contributor profile from IPFS. 

I think at least we should also add the kredits contributor id to the proposal ipfs data to be able to get to the contributor in the contract from the ipfs proposal data.

File: tests/unit/models/proposal-test.js (hunk 5 above)
Comment (fsmanuel) [reply tests/unit/models/proposal-test.js]: Yes. I'll add it in `addContributor` and `addProposal` and use `v4` to do it. I think it's cleaner to validate before we save instead of while we serialize.

File: app/services/kredits.js (hunk 4 above)
Comment (skddc) [reply app/services/kredits.js]: Cool.

File: tests/unit/models/proposal-test.js (hunk 5 above)
Comment (skddc) [reply tests/unit/models/proposal-test.js]: Sounds good.

**Comment by bumi** [comment]
yep, it uses the new contracts and the new truffle setup. That makes it way easier to develop and manage the contracts. Right now those are in: https://github.com/67P/truffle-kredits - but should replace the kredits-contract package.
Truffle has made great improvements in the last months and the ganache-cli is a great local network simulator and can easily be installed with npm. 

There is setup information here: 
https://github.com/67P/kredits-web#working-with-locally-deployed-contracts

Basically it is: 
* run and seed ganache (the local network simulator) (in truffle-kredits)
* run an offline ipfs daemon
* link the package
* run kredits-web with ember
